### PR TITLE
lncli: live table-based payment updates

### PR DESCRIPTION
**Pull request by joostjager**
Level up the cli user-experience by presenting a nice table instead of a stream of long intermediate payment state json.

lncli-live

### CODE REVIEW

File: cmd/lncli/cmd_pay.go [hunk 1]
```diff
@@ -467,20 +454,113 @@ func trackPayment(ctx *cli.Context) error {
...
 	for {
-		status, err := stream.Recv()
+		payment, err := stream.Recv()
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		printRespJSON(status)
+		table := formatPayment(payment)
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go:480]: I'd add an option to select format type, as some users may find it hard to parse the table (eg. on small terminals), or may want to parse it with `jq` or other tooling.

**Comment by bhandras** [comment]
This is really cool.

File: cmd/lncli/cmd_pay.go (hunk 1 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:480]: Yes, the only thing is that the output wasn't parseable by `jq`. It was a concatenation of multiple json messages.

**Comment by joostjager** [comment]
Added commit that shows node aliases instead of (truncated) pubkeys

File: cmd/lncli/cmd_pay.go (hunk 1 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: I don't know what to do with this. I think a basic question is whether `lncli` is meant to be used in automated processes or as a user interface. Given the large number of options to programmatically interface with `lnd`, I tend to see it as a user interface. The json output is more like an easy way to get the data on the screen.

Comment (bhandras) [reply cmd/lncli/cmd_pay.go]: For sure the table output is the nicest for human consumption. Maybe even easy to process from bash/python script if someone feels like it. Maybe an other option could be a log like output? Something where state changes are easy to debug with grep. wdyt?

Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Not sure if it is easy to parse the formatted, rewritten table.

This PR makes the output readable for humans which I think is enough. But it comes down to whether we want to support automated processes / bash / python / grep. We also don't deprecate things in `lncli`, so from a maintenance perspective it isn't ideal either to use it for automation.

File: cmd/lncli/cmd_pay.go [hunk 2]
```diff
@@ -523,7 +528,25 @@ func formatPayment(payment *lnrpc.Payment) string {
 
 		hops := []string{}
 		for _, h := range route.Hops {
-			hops = append(hops, fmt.Sprintf("%v", h.PubKey[:6]))
+			alias, ok := aliasCache[h.PubKey]
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:531]: Yes, it is non-unique, but so is a truncated pubkey. If you really want to have all info, you can check `ListPayments`. At least, that was my thought

File: cmd/lncli/cmd_pay.go [hunk 3]
```diff
@@ -467,20 +454,113 @@ func trackPayment(ctx *cli.Context) error {
...
+			state = fmt.Sprintf(
+				"%v @ %v",
+				htlc.Failure.Code,
+				htlc.Failure.FailureSourceIndex,
+			)
+		}
+
+		t.AppendRow([]interface{}{
+			state, attemptTimeText, resolveTimeText,
+			lastHop.AmtToForwardMsat, route.TotalFeesMsat,
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:563]: It isn't great, all those zeroes. Maybe we should do fractional sats?

File: cmd/lncli/cmd_pay.go [hunk 4]
```diff
@@ -467,20 +454,113 @@ func trackPayment(ctx *cli.Context) error {
...
+		table.ColumnConfig{Name: "CHAN_OUT", Align: text.AlignLeft,
+			AlignHeader: text.AlignLeft},
+	})
+
+	// Add all htlcs as rows.
+	createTime := time.Unix(0, payment.CreationTimeNs)
+	for _, htlc := range payment.Htlcs {
+		attemptTime := time.Unix(0, htlc.AttemptTimeNs)
+		attemptMs := attemptTime.Sub(createTime).Milliseconds()
+		attemptTimeText := fmt.Sprintf("%.3f", float64(attemptMs)/1000.0)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:515]: My thought was that seconds read best for the time scales that are common for this application.

File: cmd/lncli/cmd_pay.go [hunk 5]
```diff
@@ -467,20 +454,113 @@ func trackPayment(ctx *cli.Context) error {
...
+		)
+	}
+
+	// Render table.
+	b := &bytes.Buffer{}
+	t.SetOutputMirror(b)
+	t.Render()
+
+	// Add additional payment-level data.
+	fmt.Fprintf(b, "Payment status: %v", payment.Status)
```
Comment (cfromknecht) [inline cmd/lncli/cmd_pay.go:575]: would be nice to also log some more info here, like:
```
Destination:    lightning-merchant (or full pubkey)
Amount + Fee:   10000000 + 1010 mSat
Payment Hash:   713cb852db70a11ddf128695f0824055924531b8f56e3f31422321eb7690e254
Payment Status: SUCCEEDED, preimage: 89b4a95726520c3c3e9048da2808a33421337d24a0ebe39033d576df71444bfa
```
the fee could be updated live as the sum of the total fee for all non-failed htlc attempts

File: cmd/lncli/cmd_pay.go (hunk 1 above)
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go]: perhaps just a `-r` would suffice for being able to toggle back to the raw json output? it's true we don't recommend people building production things against lncli, it's primarily meant as a debugging tool for hoomans

imo it's nice to finally see lncli moving away from just logging json responses by default, and trying to present the data in a more easily consumable format.

Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: I can add `-r`, but do we still want that for the intermediate state updates? That was a concatenation of multiple json messages which isn't parseable as a whole. Maybe best to just do the final update in that case?

Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go]: It seems there are two variables here.
1. Whether or not to send non-terminal events, say `--block`.
2. Whether or not to format as table or json, say `-r`.

Both I think should be lncli flags, but only `--block` needs to be an rpc flag.

Turns out this also provides REST clients that don't support server-side streaming via keep-alive connections to use the streaming endpoint in the typical request/response patern. The url just needs to contain `?block=true` and only the terminal response is sent back before the connection is closed.

File: cmd/lncli/cmd_pay.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:524]: >nit: I'd use attemptDurationText

But it isn't a duration. It is really a time since the start of the payment.

File: cmd/lncli/cmd_pay.go [hunk 6]
```diff
@@ -467,20 +454,113 @@ func trackPayment(ctx *cli.Context) error {
...
+		attemptTime := time.Unix(0, htlc.AttemptTimeNs)
+		attemptMs := attemptTime.Sub(createTime).Milliseconds()
+		attemptTimeText := fmt.Sprintf("%.3f", float64(attemptMs)/1000.0)
+
+		resolveTimeText := "-"
+		if htlc.ResolveTimeNs != 0 {
+			resolveTime := time.Unix(0, htlc.ResolveTimeNs)
+			resolveTimeNs := resolveTime.Sub(createTime).Milliseconds()
+			resolveTimeText = fmt.Sprintf(
+				"%.3f", float64(resolveTimeNs)/1000.0,
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:531]: Discussion above

File: cmd/lncli/cmd_pay.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bhandras) [reply cmd/lncli/cmd_pay.go:524]: It's a bit confusing to read as time usually refers to a point on the timeline, while a time difference (here `attemptTime.Sub(createTime)`) is usually referred to as length or duration.

File: cmd/lncli/cmd_pay.go (hunk 1 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Thanks for the reminder about how to deal with non-streaming rest. Added the `NoInflightUpdates` flag. I think it is more descriptive than `Block`. This option already existed in `lncli` as `--show_inflight`. Didn't change that. 

Added the option to output raw json with `--json``.

File: cmd/lncli/cmd_pay.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:524]: Added nil check also for attempt time

File: cmd/lncli/cmd_pay.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:572]: Changed to fractional sats, but only if it isn't a whole sats amount

File: cmd/lncli/cmd_pay.go (hunk 5 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:651]: Added except for the destination. Would need to re-decode the payment request for that. Or inspect any of the routes last hops, but there may be none.

This also becomes more of a ux discussion. I am fine with leaving it as is now. The most important thing, readable htlcs, is there. Further improvements can be made for 0.11, also depending on what users want to see.

File: cmd/lncli/cmd_pay.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Time is always a difference. Either since the 1970, the year 0, big bang... :)

File: cmd/lncli/cmd_pay.go [hunk 7]
```diff
@@ -459,7 +455,8 @@ func trackPayment(ctx *cli.Context) error {
 	}
 
 	req := &routerrpc.TrackPaymentRequest{
-		PaymentHash: hash,
+		PaymentHash:     hash,
+		InflightUpdates: true,
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go]: Isn't this supposed to be `NoInflightUpdates` ?

File: lnrpc/routerrpc/router.proto [hunk 8]
```diff
@@ -120,11 +120,23 @@ message SendPaymentRequest {
...
 
 message TrackPaymentRequest {
     /// The hash of the payment to look up.
     bytes payment_hash = 1;
+
+    /**
+    If set, only the final payment update is streamed back. Intermediate updates
+    that show which htlcs are still in flight are suppressed.
+    */
+    bool no_inflight_updates = 2;
```
Comment (bhandras) [inline lnrpc/routerrpc/router.proto:139]: ... Or this `inflight_updates`

File: cmd/lncli/cmd_pay.go (hunk 7 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: It says comment is outdated. In the current revision, it uses `NoInflightUpdates: false` (implicitly)

File: lnrpc/routerrpc/router.proto (hunk 8 above)
Comment (joostjager) [reply lnrpc/routerrpc/router.proto:139]: The reason for doing `no_inflight_updates` is backwards compatibility. To not break even more on the rpc interface

File: cmd/lncli/cmd_pay.go (hunk 7 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Ok, turned out to be a non-building earlier commit. Fixed now

File: lnrpc/routerrpc/router.proto (hunk 8 above)
Comment (bhandras) [reply lnrpc/routerrpc/router.proto:139]: Yeah what I meant is that the commit was stale, now it's ok.

File: cmd/lncli/cmd_pay.go [hunk 9]
```diff
@@ -458,23 +464,200 @@ func trackPayment(ctx *cli.Context) error {
...
+		}
+	}
+
+	// Render table.
+	b := &bytes.Buffer{}
+	t.SetOutputMirror(b)
+	t.Render()
+
+	// Add additional payment-level data.
+
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go:647]: nit: extra new line

File: cmd/lncli/cmd_pay.go [hunk 10]
```diff
@@ -458,23 +464,200 @@ func trackPayment(ctx *cli.Context) error {
...
 
-	stream, err := client.TrackPayment(context.Background(), req)
+	stream, err := routerClient.TrackPayment(context.Background(), req)
 	if err != nil {
 		return err
 	}
 
+	client := lnrpc.NewLightningClient(conn)
+	_, err = printLivePayment(stream, client, ctx.Bool(jsonFlag.Name))
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:474]: fixed

File: cmd/lncli/cmd_pay.go [hunk 11]
```diff
@@ -458,23 +464,200 @@ func trackPayment(ctx *cli.Context) error {
...
+		return err
+	}
+
+	return nil
+}
+
+// printLivePayment uses advanced terminal control codes to rewrite the output
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:481]: updated

File: cmd/lncli/cmd_pay.go [hunk 12]
```diff
@@ -458,23 +464,200 @@ func trackPayment(ctx *cli.Context) error {
...
+}
+
+func newAliasCache(client lnrpc.LightningClient) *aliasCache {
+	return &aliasCache{
+		client: client,
+		cache:  make(map[string]string),
+	}
+}
+
+// get returns a node alias either from cache or freshly request from lnd.
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: fixed

File: cmd/lncli/cmd_pay.go (hunk 9 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: fixed

File: cmd/lncli/cmd_pay.go (hunk 5 above)
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go:650]: > Would need to re-decode the payment request for that. Or inspect any of the routes last hops, but there may be none.

We already [decode the invoice when there is one](https://github.com/lightningnetwork/lnd/blob/master/cmd/lncli/cmd_pay.go#L361), is there a way for no destination to be specified when there is no invoice? Or is there another reason we'd need to re-decode?

> This also becomes more of a ux discussion. 

This is true. Only reason I pointed these out here is because it seemed logical to at least print info we already do during confirmation, but it's not a blocker.

File: cmd/lncli/cmd_pay.go [hunk 13]
```diff
@@ -59,6 +64,11 @@ var (
 			"used",
 		Value: 1,
 	}
+
+	jsonFlag = cli.BoolFlag{
+		Name:  "json",
+		Usage: "if set, payment updates are printed as json messages",
```
Comment (cfromknecht) [inline cmd/lncli/cmd_pay.go:70]: nit: could add `-j` as the short hand

File: cmd/lncli/cmd_pay.go [hunk 14]
```diff
@@ -458,23 +464,199 @@ func trackPayment(ctx *cli.Context) error {
...
+}
+
+// formatMsat formats msat amounts as fractional sats.
+func formatMsat(amt int64) string {
+	return strconv.FormatFloat(float64(amt)/1000.0, 'f', -1, 64)
+}
+
+// formatPayment formats the payment state as an ascii table.
+func formatPayment(payment *lnrpc.Payment, aliases *aliasCache) string {
+	t := table.NewWriter()
```
Comment (cfromknecht) [inline cmd/lncli/cmd_pay.go:576]: personal preference: the `table.StyleLight` looked the cleanest to me

Comment (joostjager) [reply cmd/lncli/cmd_pay.go:576]: I tried that, but it messed up my terminal

File: cmd/lncli/cmd_pay.go (hunk 13 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:70]: Hmm, I am not even convinced that we need json output, let alone a short hand. 

I think it would help to decide at some point what we want `lncli` to be. A text-mode user interface or also a tool that we promote for usage in bash scripts.

File: cmd/lncli/cmd_pay.go (hunk 5 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:650]: If there was a confirmation step, the info is also printed above the table. If there was no confirmation step, the info is on the command line. Only when `payinvoice -f` is used, the info is only shown in the htlc rows. 

I don't want to say that this isn't a good suggestion. But if more adaptations and re-review makes this miss the cut, I'd rather leave it as is.

File: cmd/lncli/cmd_pay.go (hunk 14 above)
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go:576]: weird, i tried it locally and it worked for me. oh well

File: cmd/lncli/cmd_pay.go (hunk 13 above)
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go:70]: > A text-mode user interface or also a tool that we promote for usage in bash scripts.

I don't think the distinction needs to be that binary. It can have an improved interface for regular tasks while still provided access to the raw data for debugging.

File: cmd/lncli/cmd_pay.go (hunk 1 above)
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go]: > This option already existed in lncli as --show_inflight. Didn't change that.

Wasn't this option added in this release? If wanted to change it now is the time.

File: cmd/lncli/cmd_pay.go (hunk 5 above)
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go:650]: sgtm, we can leave as is and revisit later

File: cmd/lncli/cmd_pay.go (hunk 1 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Yes, I can rename it. But to what? I agree that `show_inflight` isn't ideal. We could mirror the rpc field, `--inflight_updates`. I prefer that over `--block`.

File: cmd/lncli/cmd_pay.go (hunk 13 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:70]: Yes, that is true. Maybe it is indeed good for the most fundamental command of all, `sendpayment`, to have a few options.

File: cmd/lncli/cmd_pay.go (hunk 14 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go:576]: I don't know what the reason is that it doesn't work. Maybe for the next release we can test a bit wider on different OSes. Who knows how Windows renders the fancy table?

**Comment by Roasbeef** [comment]
Committed as `6ffdf7ef1fa080844af9fced3a57e54d0ee22729`.

File: cmd/lncli/cmd_pay.go (hunk 1 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Renamed to `--inflight_updates`, in line with the rpc field. But always showing inflight updates for the table view

**Comment by joostjager** [comment]
Removed `uilive` dependency which broke the build on illumos and solaris. Turned out to be very easy to rewrite the terminal.

**Comment by joostjager** [comment]
Tacked on commit to rename the mpp parameter to `max_parts`, to align with the P in mPp.

**Comment by joostjager** [comment]
Lightning-rfc also consistently refers to it as `multi-part payments`

https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md#basic-multi-part-payments

File: cmd/lncli/cmd_pay.go [hunk 15]
```diff
@@ -672,3 +855,14 @@ func sendToRouteRequest(ctx *cli.Context, req *lnrpc.SendToRouteRequest) error {
 
 	return nil
 }
+
+// ESC is the ASCII code for escape character
+const ESC = 27
+
+var clear = fmt.Sprintf("%c[%dA%c[2K", ESC, 1, ESC)
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go:862]: Wouldn't something like `"\033[2K"` work?

Comment (bhandras) [reply cmd/lncli/cmd_pay.go:862]: Also please add a comment what this escape code does.

**Comment by joostjager** [comment]
Todo: test on windows

**Comment by joostjager** [comment]
Windows failed the escape sequence test, forced json.

File: cmd/lncli/cmd_pay.go (hunk 15 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Comment added, but it was explained already right below.

I don't know if other sequences work. Is there an advantage to using something else? I am no expert on this.
